hadoop-common-project/hadoop-common/src/main/java/org/apache/hadoop/util/CheckedBiFunction.java

```diff
@@ -0,0 +1,29 @@
+/**
+ * Licensed to the Apache Software Foundation (ASF) under one
+ * or more contributor license agreements.  See the NOTICE file
+ * distributed with this work for additional information
+ * regarding copyright ownership.  The ASF licenses this file
+ * to you under the Apache License, Version 2.0 (the
+ * "License"); you may not use this file except in compliance
+ * with the License.  You may obtain a copy of the License at
+ *
+ *     http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+
+package org.apache.hadoop.util;
+
+import java.io.IOException;
+
+/**
+ * Defines a functional interface having two inputs which throws IOException.
+ */
+@FunctionalInterface
+public interface CheckedBiFunction<LEFT, RIGHT, THROWABLE extends IOException> {
```

Comment:
Cute! I never knew you could do that with templates and exceptions!

1. can you put into org.apache.hadoop.fs.impl where the other internal-for-fs-only lambda stuff is going.
1. be advised that for backports to branch 2 we will have to make things compile on Java 8. Mostly this is just using the IDE to convert things to callables. Doesn't mean they shouldn't be used, only that once you get sufficiently advanced things become unbackportable. This patch looks fine

Reply:
Actually CheckedBiFunction is already available in other place (HDDS) - I feel these interfaces/classes would be better to be moved into common module when Hadoop could forget about JDK7.

Reply:
yeah, they should. I'm doing in new code that I know isn't going to be backportable into jdk7, not now we have to worry about jdk11